cmd/telegraf/main.go

```diff
@@ -0,0 +1,352 @@
+			Name:  "processor-filter",
+			Usage: "filter the processors to enable, separator is ':'",
+		},
+	}
+
+	extraFlags := append(pluginFilterFlags, cliFlags()...)
+
+	app := &cli.App{
+		Name:   "Telegraf",
+		Usage:  "The plugin-driven server agent for collecting & reporting metrics.",
+		Writer: outputBuffer,
+		Flags: append(
+			[]cli.Flag{
+				// String slice flags
+				&cli.StringSliceFlag{
+					Name:  "config",
+					Usage: "configuration file to load",
+				},
+				&cli.StringSliceFlag{
+					Name:  "config-directory",
+					Usage: "directory containing additional *.conf files",
+				},
+				// Int flags
+				&cli.IntFlag{
+					Name:  "test-wait",
+					Usage: "wait up to this many seconds for service inputs to complete in test mode",
+				},
+				//
+				// String flags
+				&cli.StringFlag{
+					Name:  "usage",
+					Usage: "print usage for a plugin, ie, 'telegraf --usage mysql'",
+				},
+				&cli.StringFlag{
+					Name:  "pprof-addr",
+					Usage: "pprof host/IP and port to listen on (e.g. 'localhost:6060')",
+				},
+				&cli.StringFlag{
+					Name:  "watch-config",
+					Usage: "monitoring config changes [notify, poll]",
+				},
+				&cli.StringFlag{
+					Name:  "pidfile",
+					Usage: "file to write our pid to",
+				},
+				//
+				// Bool flags
+				&cli.BoolFlag{
+					Name:  "once",
+					Usage: "run one gather and exit",
+				},
+				&cli.BoolFlag{
+					Name:  "debug",
+					Usage: "turn on debug logging",
+				},
+				&cli.BoolFlag{
+					Name:  "quiet",
+					Usage: "run in quiet mode",
+				},
+				&cli.BoolFlag{
+					Name:  "test",
+					Usage: "enable test mode: gather metrics, print them out, and exit. Note: Test mode only runs inputs, not processors, aggregators, or outputs",
+				},
+				// TODO: Change "deprecation-list, input-list, output-list" flags to become a subcommand "list" that takes
+				// "input,output,aggregator,processor, deprecated" as parameters
+				&cli.BoolFlag{
+					Name:  "deprecation-list",
+					Usage: "print all deprecated plugins or plugin options",
+				},
+				&cli.BoolFlag{
+					Name:  "input-list",
+					Usage: "print available input plugins",
+				},
+				&cli.BoolFlag{
+					Name:  "output-list",
+					Usage: "print available output plugins",
+				},
+				//
+				// !!! The following flags are DEPRECATED !!!
+				// Already covered with the subcommand `./telegraf version`
+				&cli.BoolFlag{
+					Name:  "version",
+					Usage: "DEPRECATED: display the version and exit",
+				},
+				// Already covered with the subcommand `./telegraf config`
+				&cli.BoolFlag{
+					Name:  "sample-config",
+					Usage: "DEPRECATED: print out full sample configuration",
+				},
+				// Using execd plugin to add external plugins is preffered (less size impact, easier for end user)
+				&cli.StringFlag{
+					Name:  "plugin-directory",
+					Usage: "DEPRECATED: path to directory containing external plugins",
+				},
+				// !!!
+			}, extraFlags...),
+		Action: func(cCtx *cli.Context) error {
```

Comment:
I like how urfave/cli uses structs nested in cli.App to configure cli options (flags and commands) but I think including funcs in the middle of the configuration structs (the `Action` fields) makes it harder to read when the func is long like it is here. The func's indentation is pretty deep here, five levels if I'm counting right.

What do you think about moving this func to be defined before the cli.App around line 85, then refer to it here by name? I think it will be one or two less indentation levels which should be a bit easier to read.

Reply:
Seems like a good idea to me, less indentation is definitely better. I've moved the function to line 85